dace/codegen/targets/fpga.py

```diff
@@ -67,6 +68,186 @@ def is_fpga_kernel(sdfg, state):
             return False
     return True
 
+def is_hbm_array(array: dt.Data):
+    """
+    :return: True if this array is placed on HBM
+    """
+    if (isinstance(array, dt.Array)
+            and array.storage == dtypes.StorageType.FPGA_Global):
+        res = parse_location_bank(array)
+        return res is not None and res[0] == "HBM"
+    else:
+        return False
+
+
+def is_fpga_array(array: dt.Data):
+    """
+    :return: True if this array is placed on FPGA memory
+    """
+    return isinstance(array, dt.Array) and array.storage in _FPGA_STORAGE_TYPES
+
+
+def iterate_hbm_multibank_arrays(array_name: str, array: dt.Array, sdfg: SDFG):
+    """
+    Small helper function that iterates over the bank indices
+    if the provided array is spanned across multiple HBM banks.
+    Otherwise just returns 0 once.
+    """
+    res = parse_location_bank(array)
+    if res is not None:
+        bank_type, bank_place = res
+        if (bank_type == "HBM"):
+            low, high = get_multibank_ranges_from_subset(bank_place, sdfg)
+            for i in range(high - low):
+                yield i
+        else:
+            yield 0
+    else:
+        yield 0
+
+
+def modify_distributed_subset(subset: Union[subsets.Subset, Iterable],
+                              change: int):
+    """
+    Modifies the first index of :param subset: (the one used for distributed subsets).
+    :param subset: is deepcopied before any modification to it is done.
+    :param change: the first index is set to this value, unless it's (-1) in which case
+        the first index is completly removed
+    """
+    cps = copy.deepcopy(subset)
+    if isinstance(subset, subsets.Subset):
+        if change == -1:
+            cps.pop([0])
+        else:
+            cps[0] = (change, change, 1)
+    elif isinstance(subset, Iterable):
+        cps = list(cps)  # Works for any iterable type
+        if change == -1:
+            cps = cps[1:]
+        else:
+            cps[0] = change
```

Comment:
This seems inconsistent, in the first part you change it to a tuple `(change, change, 1)`, in the second to `change`? Probably only one is correct?

You also don't really need the `isinstance(subset, Iterable)`, I would have an `else` and convert to a list directly (it'll fail anyway if it doesn't work). Whatever you prefer.

Actually, couldn't this whole thing be replaced with just (with no `isinstance` check):
```python
cps = list(subset)
if change == -1:
  return cps[1:]
else:
  cps[0] = (change, change, 1)
  return cps
```

Reply:
just assigning (change, change, 1) always will not work. The function is used in multiple contexts, to update shapes as well as subsets. Probably the clean way to go would be to actually have two different functions.

Reply:
Ok, instead I know completly threw out the support for shapes, since this is anyway only used in one place. So now this actually does except Subset as type in all cases.